File: frontend/mobile/lib/features/call/services/webrtc_call_service.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:vnalo_mobile/services/socket_service.dart';

class WebRtcCallService extends ChangeNotifier {
  static const int _defaultRingTimeoutSeconds = int.fromEnvironment(
    'CALL_RING_TIMEOUT_SECONDS',
    defaultValue: 38,
  );

  final SocketService _socketService;
  final String conversationId;
  final String callId;
  final String currentUserId;
  final String peerUserId;
  final bool audioOnly;
  final bool isCaller;
  final int ringTimeoutSeconds;
  final Map<String, dynamic>? initialSdp;

  RTCPeerConnection? _peerConnection;
  MediaStream? _localStream;
  MediaStream? _remoteStream;
  StreamSubscription<Map<String, dynamic>>? _signalSubscription;
  final List<RTCIceCandidate> _pendingCandidates = [];
  Timer? _ringTimeoutTimer;

  bool _isInitializing = false;
  bool _isConnected = false;
  bool _isEnded = false;
  bool _isMicrophoneEnabled = true;
  bool _isSpeakerOn = true;
  bool _isCameraEnabled = true;
  bool _isUsingFrontCamera = true;
  String? _errorMessage;
  DateTime? _connectedAt;
  bool _hasRemoteDescription = false;
  String? _lastEndReason;
  bool _isAccepted = false;

  WebRtcCallService({
    required SocketService socketService,
    required this.conversationId,
    required this.callId,
    required this.currentUserId,
    required this.peerUserId,
    required this.audioOnly,
    required this.isCaller,
    this.initialSdp,
    int? ringTimeoutSeconds,
  }) : ringTimeoutSeconds =
           (ringTimeoutSeconds == null || ringTimeoutSeconds <= 0)
               ? _defaultRingTimeoutSeconds
               : ringTimeoutSeconds,
       _socketService = socketService;

  MediaStream? get localStream => _localStream;
  MediaStream? get remoteStream => _remoteStream;
  bool get isInitializing => _isInitializing;
  bool get isConnected => _isConnected;
  bool get isEnded => _isEnded;
  bool get isMicrophoneEnabled => _isMicrophoneEnabled;
  bool get isSpeakerOn => _isSpeakerOn;
  bool get isCameraEnabled => _isCameraEnabled;
  bool get isUsingFrontCamera => _isUsingFrontCamera;
  String? get errorMessage => _errorMessage;
  DateTime? get connectedAt => _connectedAt;
  String? get lastEndReason => _lastEndReason;
  bool get isAccepted => _isAccepted;

  String get _logPrefix =>
      '[WebRtcCallService][callId=$callId][conv=$conversationId][peer=$peerUserId]';

  Future<void> initialize() async {
    if (_isInitializing || _peerConnection != null) return;
    _isInitializing = true;
    _errorMessage = null;
    _isCameraEnabled = !audioOnly;
    notifyListeners();

    _signalSubscription = _socketService.onCallSignal.listen(_onSignalEvent);
    debugPrint(
      '$_logPrefix initialize() start isCaller=$isCaller audioOnly=$audioOnly timeout=$ringTimeoutSeconds hasInitialSdp=${initialSdp != null}',
    );

    try {
      // Proactive permission check
      final micStatus = await Permission.microphone.request();
      if (micStatus.isDenied || micStatus.isPermanentlyDenied) {
        throw 'Quyền micro bị từ chối';
      }
      if (!audioOnly) {
        final camStatus = await Permission.camera.request();
        if (camStatus.isDenied || camStatus.isPermanentlyDenied) {
          throw 'Quyền camera bị từ chối';
        }
      }

      final configuration = {
        'iceServers': [
          {'urls': 'stun:stun.l.google.com:19302'},
          {'urls': 'stun:stun1.l.google.com:19302'},
          // Placeholder for TURN servers - requires credentials for production
          // {
          //   'urls': 'turn:your-turn-server.com:3478',
          //   'username': 'user',
          //   'credential': 'password'
          // },
        ],
        'sdpSemantics': 'unified-plan',
      };

      _peerConnection = await createPeerConnection(configuration);
      debugPrint('$_logPrefix peerConnection created');
      _registerPeerCallbacks();

      if (isCaller) {
        _isAccepted = true; 
        await _openLocalMedia();
        await Helper.setSpeakerphoneOn(_isSpeakerOn);
        await _createAndSendOffer();
        _startRingTimeoutCountdown();
      } else {
        // Callee: Immediately process initial offer if provided
        if (initialSdp != null) {
          await _handleOffer(initialSdp!);
        }
        await Helper.setSpeakerphoneOn(_isSpeakerOn);
      }
    } catch (e) {
      _errorMessage = _mapInitError(e);
      debugPrint('$_logPrefix initialize() failed error=$_errorMessage');
    } finally {
      _isInitializing = false;
      notifyListeners();
    }
  }

  String _mapInitError(Object error) {
    final text = error.toString().toLowerCase();
    if (text.contains('permission') || text.contains('notallowed') || text.contains('từ chối')) {
      return 'Không có quyền truy cập camera/micro. Vui lòng cấp quyền rồi thử lại.';
    }
    return 'Không thể khởi tạo cuộc gọi: $error';
  }

  void _registerPeerCallbacks() {
    _peerConnection?.onTrack = (RTCTrackEvent event) {
      if (event.streams.isEmpty) return;
      _remoteStream = event.streams.first;
      debugPrint('$_logPrefix onTrack remote stream attached');
      notifyListeners();
    };

    _peerConnection?.onIceCandidate = (RTCIceCandidate candidate) {
      if (candidate.candidate == null) return;
      _socketService.sendCallIceCandidate(
        conversationId: conversationId,
        callId: callId,
        targetUserId: peerUserId,
        senderUserId: currentUserId,
        candidate: {
          'candidate': candidate.candidate,
          'sdpMid': candidate.sdpMid,
          'sdpMLineIndex': candidate.sdpMLineIndex,
        },
      );
    };

    _peerConnection?.onConnectionState = (RTCPeerConnectionState state) {
      debugPrint('$_logPrefix onConnectionState=$state');
      if (state == RTCPeerConnectionState.RTCPeerConnectionStateConnected) {
        if (!_isConnected) {
          _isConnected = true;
          _connectedAt = DateTime.now();
          _ringTimeoutTimer?.cancel();
          notifyListeners();
        }
      } else if (state == RTCPeerConnectionState.RTCPeerConnectionStateFailed ||
                 state == RTCPeerConnectionState.RTCPeerConnectionStateDisconnected ||
                 state == RTCPeerConnectionState.RTCPeerConnectionStateClosed) {
         _isConnected = false;
         notifyListeners();
      }
    };

    _peerConnection?.onSignalingState = (RTCSignalingState state) {
      debugPrint('$_logPrefix onSignalingState=$state');
    };

    _peerConnection?.onIceConnectionState = (RTCIceConnectionState state) {
      debugPrint('$_logPrefix onIceConnectionState=$state');
      // Fallback for older WebRTC versions if onConnectionState is not consistent
      if (state == RTCIceConnectionState.RTCIceConnectionStateConnected ||
          state == RTCIceConnectionState.RTCIceConnectionStateCompleted) {
        if (!_isConnected) {
          _isConnected = true;
          _connectedAt = DateTime.now();
          _ringTimeoutTimer?.cancel();
          notifyListeners();
        }
      }
    };
  }

  Future<void> _openLocalMedia() async {
    final mediaConstraints = {
      'audio': true,
      'video':
          audioOnly
              ? false
              : {
                'facingMode': 'user',
                'width': 1280,
                'height': 720,
                'frameRate': 30,
              },
    };

    _localStream = await navigator.mediaDevices.getUserMedia(mediaConstraints);
    debugPrint(
      '$_logPrefix local media opened tracks=${_localStream?.getTracks().length ?? 0}',
    );
    final tracks = _localStream?.getTracks() ?? const <MediaStreamTrack>[];
    for (final track in tracks) {
      await _peerConnection?.addTrack(track, _localStream!);
    }
  }

  Future<void> acceptCall() async {
    if (_isAccepted || _isEnded) return;
    debugPrint('$_logPrefix acceptCall()');

    try {
      _isAccepted = true;
      notifyListeners();

      await _openLocalMedia();
      
      if (_hasRemoteDescription) {
        await _createAndSendAnswer();
      }
    } catch (e) {
      _errorMessage = 'Không thể chấp nhận cuộc gọi: $e';
      debugPrint('$_logPrefix acceptCall() failed error=$_errorMessage');
      notifyListeners();
    }
  }

  Future<void> _createAndSendAnswer() async {
    final pc = _peerConnection;
    if (pc == null) return;

    final answer = await pc.createAnswer({
      'offerToReceiveAudio': 1,
      'offerToReceiveVideo': audioOnly ? 0 : 1,
    });
    await pc.setLocalDescription(answer);
    debugPrint('$_logPrefix local answer created and set');

    _socketService.sendCallAnswer(
      conversationId: conversationId,
      callId: callId,
      targetUserId: peerUserId,
      senderUserId: currentUserId,
      sdp: {'type': answer.type, 'sdp': answer.sdp},
    );
  }

  Future<void> _createAndSendOffer() async {
    final pc = _peerConnection;
    if (pc == null) return;

    final offer = await pc.createOffer({
      'offerToReceiveAudio': 1,
      'offerToReceiveVideo': audioOnly ? 0 : 1,
    });
    await pc.setLocalDescription(offer);
    debugPrint('$_logPrefix local offer created and set');

    _socketService.sendCallOffer(
      conversationId: conversationId,
      callId: callId,
      targetUserId: peerUserId,
      senderUserId: currentUserId,
      audioOnly: audioOnly,
      sdp: {'type': offer.type, 'sdp': offer.sdp},
    );
  }

  void _startRingTimeoutCountdown() {
    _ringTimeoutTimer?.cancel();
    if (!isCaller || ringTimeoutSeconds <= 0) return;

    _ringTimeoutTimer = Timer(Duration(seconds: ringTimeoutSeconds), () async {
      if (_isEnded || _isConnected) return;
      _errorMessage =
          'Người nhận chưa bắt máy sau $ringTimeoutSeconds giây. Cuộc gọi đã tự động kết thúc.';
      debugPrint(
        '$_logPrefix ring timeout reached after $ringTimeoutSeconds seconds',
      );
      notifyListeners();
      await endCall(reason: 'no-answer-timeout');
    });
  }

  Future<void> _onSignalEvent(Map<String, dynamic> signal) async {
    if (_isEnded) return;

    final signalConversationId = signal['conversationId']?.toString();
    final signalCallId = signal['callId']?.toString();
    if (signalConversationId != conversationId || signalCallId != callId) {
      return;
    }

    final senderUserId = _extractUserId(signal, ['senderUserId', 'senderId', 'fromUserId']);
    if (senderUserId != null && senderUserId.isNotEmpty && senderUserId != peerUserId) {
      return;
    }

    final type = signal['type']?.toString();
    if (type == null || type.isEmpty) return;
    
    // De-duplication check: if we already have the remote description for this type, skip.
    if (type == 'offer' && _hasRemoteDescription && !isCaller) {
      debugPrint('$_logPrefix skipping duplicate offer signal');
      return;
    }

    debugPrint('$_logPrefix received signal type=$type');

    try {
      switch (type) {
        case 'offer':
          if (!isCaller) {
            await _handleOffer(signal);
          }
          break;
        case 'answer':
          if (isCaller) {
            await _handleAnswer(signal);
          }
          break;
        case 'ice-candidate':
          await _handleIceCandidate(signal);
          break;
        case 'end':
          await endCall(
            notifyPeer: false,
            reason: signal['reason']?.toString() ?? 'remote-ended',
          );
          break;
      }
    } catch (e) {
      debugPrint('$_logPrefix error processing signal type=$type: $e');
    }
  }

  String? _extractUserId(Map<String, dynamic> signal, List<String> keys) {
    for (final key in keys) {
      final value = signal[key]?.toString();
      if (value != null && value.isNotEmpty) {
        return value;
      }
    }
    return null;
  }

  Future<void> _handleOffer(Map<String, dynamic> signal) async {
    final pc = _peerConnection;
    if (pc == null) return;
    final sdpMap = Map<String, dynamic>.from(signal['sdp'] ?? {});

    final offer = RTCSessionDescription(
      sdpMap['sdp']?.toString() ?? '',
      sdpMap['type']?.toString() ?? 'offer',
    );

    await pc.setRemoteDescription(offer);
    debugPrint('$_logPrefix remote offer set');
    _hasRemoteDescription = true;
    await _flushPendingCandidates();

    if (_isAccepted) {
      await _createAndSendAnswer();
    }
  }

  Future<void> _handleAnswer(Map<String, dynamic> signal) async {
    final pc = _peerConnection;
    if (pc == null) return;
    final sdpMap = Map<String, dynamic>.from(signal['sdp'] ?? {});

    final answer = RTCSessionDescription(
      sdpMap['sdp']?.toString() ?? '',
      sdpMap['type']?.toString() ?? 'answer',
    );

    // Guard against setting duplicate answers or racing signals
    if (pc.signalingState == RTCSignalingState.RTCSignalingStateStable) {
      debugPrint('$_logPrefix ignoring answer as pc is already stable');
      return;
    }

    await pc.setRemoteDescription(answer);
    debugPrint('$_logPrefix remote answer set');
    _hasRemoteDescription = true;
    await _flushPendingCandidates();
  }

  Future<void> _handleIceCandidate(Map<String, dynamic> signal) async {
    final pc = _peerConnection;
    if (pc == null) return;

    final candidateMap = Map<String, dynamic>.from(signal['candidate'] ?? {});
    final candidateString = candidateMap['candidate']?.toString();
    if (candidateString == null) return;

    final candidate = RTCIceCandidate(
      candidateString,
      candidateMap['sdpMid']?.toString(),
      candidateMap['sdpMLineIndex'] is int
          ? candidateMap['sdpMLineIndex'] as int
          : int.tryParse(candidateMap['sdpMLineIndex']?.toString() ?? ''),
    );

    if (!_hasRemoteDescription) {
      debugPrint('$_logPrefix queue ICE candidate before remote description');
      _pendingCandidates.add(candidate);
      return;
    }

    debugPrint('$_logPrefix add ICE candidate immediately');
    await pc.addCandidate(candidate);
  }

  Future<void> _flushPendingCandidates() async {
    if (!_hasRemoteDescription || _pendingCandidates.isEmpty) return;
    final pc = _peerConnection;
    if (pc == null) return;

    for (final candidate in List<RTCIceCandidate>.from(_pendingCandidates)) {
      await pc.addCandidate(candidate);
    }
    debugPrint('$_logPrefix flushed ${_pendingCandidates.length} queued ICE candidates');
    _pendingCandidates.clear();
  }

  Future<void> toggleMicrophone() async {
    final audioTracks = _localStream?.getAudioTracks() ?? const <MediaStreamTrack>[];
    if (audioTracks.isEmpty) return;

    _isMicrophoneEnabled = !_isMicrophoneEnabled;
    for (final track in audioTracks) {
      track.enabled = _isMicrophoneEnabled;
    }
    notifyListeners();
  }

  Future<void> toggleSpeaker() async {
    _isSpeakerOn = !_isSpeakerOn;
    await Helper.setSpeakerphoneOn(_isSpeakerOn);
    notifyListeners();
  }

  Future<void> toggleCamera() async {
    if (audioOnly) return;
    final videoTracks = _localStream?.getVideoTracks() ?? const <MediaStreamTrack>[];
    if (videoTracks.isEmpty) return;

    _isCameraEnabled = !_isCameraEnabled;
    for (final track in videoTracks) {
      track.enabled = _isCameraEnabled;
    }
    notifyListeners();
  }

  Future<void> switchCamera() async {
    if (audioOnly) return;
    final videoTracks = _localStream?.getVideoTracks() ?? const <MediaStreamTrack>[];
    if (videoTracks.isEmpty) return;

    await Helper.switchCamera(videoTracks.first);
    _isUsingFrontCamera = !_isUsingFrontCamera;
    notifyListeners();
  }

  Future<void> endCall({
    bool notifyPeer = true,
    String reason = 'hangup',
  }) async {
    if (_isEnded) return;

    debugPrint('$_logPrefix endCall notifyPeer=$notifyPeer reason=$reason');

    _isEnded = true;
    _isConnected = false;
    _lastEndReason = reason;
    _ringTimeoutTimer?.cancel();
    _ringTimeoutTimer = null;

    if (notifyPeer) {
      _socketService.endCall(
        conversationId: conversationId,
        callId: callId,
        targetUserId: peerUserId,
        senderUserId: currentUserId,
        reason: reason,
      );
    }

    await _signalSubscription?.cancel();
    _signalSubscription = null;

    try {
      await _peerConnection?.close();
    } catch (_) {}
    _peerConnection = null;
    _pendingCandidates.clear();
    _hasRemoteDescription = false;

    await _disposeStreams();
    notifyListeners();
  }

  Future<void> _disposeStreams() async {
    final localTracks = _localStream?.getTracks() ?? const <MediaStreamTrack>[];
    for (final track in localTracks) {
      track.stop();
    }
    await _localStream?.dispose();
    _localStream = null;

    final remoteTracks = _remoteStream?.getTracks() ?? const <MediaStreamTrack>[];
    for (final track in remoteTracks) {
      track.stop();
    }
    await _remoteStream?.dispose();
    _remoteStream = null;
  }

  @override
  void dispose() {
    _ringTimeoutTimer?.cancel();
    unawaited(endCall(notifyPeer: false));
    super.dispose();
  }
}

File: frontend/web/src/features/chat/webrtcCallService.ts
import { Socket } from 'socket.io-client'

export type CallOutcome = 'completed' | 'canceled' | 'missed'

export interface WebRTCCallState {
  pc: RTCPeerConnection | null
  localStream: MediaStream | null
  remoteStream: MediaStream | null
  isConnected: boolean
  isEnded: boolean
  isMicOn: boolean
  isCameraOn: boolean
  isRemoteMicOn: boolean
  isRemoteCameraOn: boolean
  startedAt?: number
  hasRemoteDescription: boolean
  pendingCandidates: RTCIceCandidate[]
  error: string | null
}

export class WebRtcCallService {
  private state: WebRTCCallState = {
    pc: null,
    localStream: null,
    remoteStream: null,
    isConnected: false,
    isEnded: false,
    isMicOn: true,
    isCameraOn: true,
    isRemoteMicOn: true,
    isRemoteCameraOn: true,
    pendingCandidates: [],
    hasRemoteDescription: false,
    error: null,
  }

  private resetInternalState() {
    this.state.hasRemoteDescription = false
    this.state.pendingCandidates = []
    this.state.isConnected = false
    this.state.isEnded = false
    this.state.error = null
  }

  private socket: Socket | null = null
  private conversationId: string = ''
  private callId: string = ''
  private currentUserId: string = ''
  private peerUserId: string = ''
  private audioOnly: boolean = false
  private isCaller: boolean = false
  private onStateChange: (state: WebRTCCallState) => void

  private ringTimeoutTimer: ReturnType<typeof setTimeout> | null = null
  private processedIceCandidates: Set<string> = new Set()
  private readonly RING_TIMEOUT_MS = 38000

  constructor(onStateChange: (state: WebRTCCallState) => void) {
    this.onStateChange = onStateChange
  }

  private updateState(patch: Partial<WebRTCCallState>) {
    this.state = { ...this.state, ...patch }
    this.onStateChange(this.state)
  }

  async initialize(params: {
    socket: Socket
    conversationId: string
    callId: string
    currentUserId: string
    peerUserId: string
    audioOnly: boolean
    isCaller: boolean
    initialSdp?: RTCSessionDescriptionInit
  }) {
    console.log('[WebRTC] Initializing service', params)
    this.socket = params.socket
    this.conversationId = params.conversationId
    this.callId = params.callId
    this.currentUserId = params.currentUserId
    this.peerUserId = params.peerUserId
    this.audioOnly = params.audioOnly
    this.isCaller = params.isCaller

    try {
      // 1. Setup PeerConnection
      const turnUrl = import.meta.env.VITE_TURN_URL;
      const turnUser = import.meta.env.VITE_TURN_USERNAME;
      const turnPass = import.meta.env.VITE_TURN_PASSWORD;

      const configuration: RTCConfiguration = {
        iceServers: [
          { urls: 'stun:stun.l.google.com:19302' },
          { urls: 'stun:stun1.l.google.com:19302' },
          { urls: 'stun:stun2.l.google.com:19302' },
          { urls: 'stun:stun3.l.google.com:19302' },
          { urls: 'stun:stun4.l.google.com:19302' },
        ],
        iceCandidatePoolSize: 10,
      };

      if (turnUrl && turnUser && turnPass) {
        console.log('[WebRTC] Adding TURN server:', turnUrl);
        configuration.iceServers?.push({
          urls: turnUrl,
          username: turnUser,
          credential: turnPass,
        });
      }

      this.resetInternalState()
      this.state.pc = new RTCPeerConnection(configuration)

      this.registerPeerEvents()

      // 2. Setup Local Media
      await this.openLocalMedia()

      // 3. Handshake logic
      if (this.isCaller) {
        await this.createOffer()
        this.startRingTimeout()
      } else if (params.initialSdp) {
        await this.handleOffer(params.initialSdp)
      }
    } catch (error) {
      console.error('[WebRTC] Initialization failed', error)
      this.endCall('failed')
    }
  }

  private registerPeerEvents() {
    const pc = this.state.pc
    if (!pc) return

    pc.onicecandidate = (event) => {
      if (event.candidate && this.socket) {
        const candidatePayload = {
          candidate: event.candidate.candidate,
          sdpMid: event.candidate.sdpMid,
          sdpMLineIndex: event.candidate.sdpMLineIndex,
        }

        const payload = {
          conversationId: this.conversationId,
          callId: this.callId,
          targetUserId: this.peerUserId,
          senderUserId: this.currentUserId,
          callerId: this.isCaller ? this.currentUserId : this.peerUserId,
          calleeId: this.isCaller ? this.peerUserId : this.currentUserId,
          roomId: this.conversationId,
          candidate: candidatePayload,
          type: 'ice-candidate',
        }

        this.socket.emit('call:ice-candidate', payload)
        this.socket.emit('call.ice-candidate', payload)
      }
    }

    pc.ontrack = (event) => {
      console.log('[WebRTC] Remote track received:', event.track.kind)
      if (event.streams && event.streams[0]) {
        this.updateState({ remoteStream: event.streams[0] })
      }

      // ──────────────────────────────────────────────────────────────────
      // KEY FIX: Listen to the track's native mute/unmute events.
      // When a remote peer toggles their camera/mic, the browser fires
      // 'mute' / 'unmute' on the corresponding MediaStreamTrack.
      // This is 100% frontend-only, no backend changes needed.
      // ──────────────────────────────────────────────────────────────────
      const track = event.track

      if (track.kind === 'video') {
        track.onmute = () => {
          console.log('[WebRTC] 🎥 Remote video track MUTED → showing avatar')
          this.updateState({ isRemoteCameraOn: false })
        }
        track.onunmute = () => {
          console.log('[WebRTC] 🎥 Remote video track UNMUTED → showing video')
          this.updateState({ isRemoteCameraOn: true })
        }
        // Set initial state based on track's current mute status
        this.updateState({ isRemoteCameraOn: !track.muted })
      }

      if (track.kind === 'audio') {
        track.onmute = () => {
          console.log('[WebRTC] 🎤 Remote audio track MUTED')
          this.updateState({ isRemoteMicOn: false })
        }
        track.onunmute = () => {
          console.log('[WebRTC] 🎤 Remote audio track UNMUTED')
          this.updateState({ isRemoteMicOn: true })
        }
        this.updateState({ isRemoteMicOn: !track.muted })
      }
    }

    pc.onconnectionstatechange = () => {
      console.log('[WebRTC] Connection state changed:', pc.connectionState)
      this.updateState({ isConnected: pc.connectionState === 'connected' })
      
      if (pc.connectionState === 'connected') {
        this.stopRingTimeout()
        this.updateState({ startedAt: Date.now(), error: null })
      } else if (pc.connectionState === 'failed') {
        console.error('[WebRTC] Connection failed. Signaling state:', pc.signalingState, 'ICE Gathering:', pc.iceGatheringState)
        this.updateState({ error: 'Kết nối mạng thất bại' })
      } else if (pc.connectionState === 'disconnected') {
        console.warn('[WebRTC] Connection disconnected')
      }
    }

    pc.onsignalingstatechange = () => {
      console.log('[WebRTC] Signaling state changed:', pc.signalingState)
    }

    pc.onicegatheringstatechange = () => {
      console.log('[WebRTC] ICE gathering state changed:', pc.iceGatheringState)
    }
  }

  public async openLocalMedia() {
    if (this.state.localStream) return this.state.localStream;
    try {
      const constraints = {
        audio: {
          echoCancellation: true,
          noiseSuppression: true,
          autoGainControl: true,
        },
        video: this.audioOnly ? false : { 
          facingMode: 'user', 
          width: { ideal: 1280 }, 
          height: { ideal: 720 } 
        },
      }

      console.log('[WebRTC] Requesting media:', constraints)

      // ── ORIGIN & BROWSER CAPABILITY CHECK ──────────────────────────
      const isSecure = location.protocol === 'https:' || location.hostname === 'localhost' || location.hostname === '127.0.0.1'
      if (!isSecure) {
        const msg = 'Vui lòng sử dụng HTTPS để truy cập micro và camera. Ví dụ: https://' + location.hostname
        console.error('[WebRTC] Insecure origin:', location.protocol, location.hostname)
        this.updateState({ error: msg })
        throw new Error(msg)
      }

      if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
        const msg = 'Trình duyệt của bạn không hỗ trợ truy cập Camera/Microphone hoặc tính năng này đã bị chặn.'
        this.updateState({ error: msg })
        throw new Error(msg)
      }

      const mediaPromise = navigator.mediaDevices.getUserMedia(constraints)
      const timeoutPromise = new Promise<never>((_, reject) =>
        setTimeout(() => reject(new Error('Yêu cầu quyền truy cập Media bị quá hạn (30s). Vui lòng nhấn "Allow" khi trình duyệt hỏi.')), 30000)
      )

      const stream = await Promise.race([mediaPromise, timeoutPromise])
      
      console.log('[WebRTC] Local media stream obtained', {
        id: stream.id,
        audio: stream.getAudioTracks().length,
        video: stream.getVideoTracks().length
      })
      this.updateState({ localStream: stream })
      
      if (this.state.pc && this.state.isConnected) {
         this.updateState({ startedAt: Date.now() })
      }

      stream.getTracks().forEach((track) => {
        if (this.state.pc) {
          this.state.pc.addTrack(track, stream)
        }
      })
    } catch (error) {
      console.error('[WebRTC] Failed to open local media', error)
      const errorMsg = error instanceof Error ? error.message : String(error)
      this.updateState({ error: `Lỗi Microphone/Camera: ${errorMsg}` })
      throw error
    }
  }

  private async createOffer() {
    const pc = this.state.pc
    if (!pc) return

    const offer = await pc.createOffer({
      offerToReceiveAudio: true,
      offerToReceiveVideo: !this.audioOnly
    })
    await pc.setLocalDescription(offer)

    const offerPayload = {
      conversationId: this.conversationId,
      callId: this.callId,
      senderUserId: this.currentUserId,
      targetUserId: this.peerUserId,
      callerId: this.currentUserId,
      calleeId: this.peerUserId,
      roomId: this.conversationId,
      mediaType: this.audioOnly ? 'voice' : 'video',
      audioOnly: this.audioOnly,
      sdp: { type: offer.type, sdp: offer.sdp },
    }

    if (this.socket) {
      this.socket.emit('call:offer', offerPayload)
      this.socket.emit('call.offer', offerPayload)

    }
  }

  async acceptCall() {
    if (this.state.isEnded || !this.state.hasRemoteDescription) {
      console.warn('[WebRTC] Cannot accept call: Ended or missing remote description')
      return
    }

    const pc = this.state.pc
    if (!pc) return

    if (pc.signalingState !== 'have-remote-offer') {
      console.warn('[WebRTC] Cannot accept call: PC signalingState is', pc.signalingState)
      return
    }

    try {
      const answer = await pc.createAnswer({
        offerToReceiveAudio: true,
        offerToReceiveVideo: !this.audioOnly
      })
      await pc.setLocalDescription(answer)

      const answerPayload = {
        conversationId: this.conversationId,
        callId: this.callId,
        senderUserId: this.currentUserId,
        targetUserId: this.peerUserId,
        callerId: this.peerUserId,
        calleeId: this.currentUserId,
        roomId: this.conversationId,
        mediaType: this.audioOnly ? 'voice' : 'video',
        sdp: { type: answer.type, sdp: answer.sdp },
      }

      if (this.socket) {
        this.socket.emit('call:answer', answerPayload)
        this.socket.emit('call.answer', answerPayload)

      }
    } catch (error) {
      console.error('[WebRTC] Failed to accept call', error)
    }
  }

  async handleOffer(offerSdp: RTCSessionDescriptionInit) {
    if (this.state.hasRemoteDescription && !this.isCaller) return

    const pc = this.state.pc
    if (!pc) return

    try {
      await pc.setRemoteDescription(new RTCSessionDescription(offerSdp))
      this.updateState({ 
        hasRemoteDescription: true,
      })
      await this.flushPendingCandidates()
    } catch (error) {
      console.error('[WebRTC] Failed to handle offer', error)
    }
  }

  async handleAnswer(answerSdp: RTCSessionDescriptionInit) {
    const pc = this.state.pc
    if (!pc) return

    if (pc.signalingState === 'stable') {
      console.warn('[WebRTC] Ignoring handleAnswer: peer connection is already stable')
      return
    }

    try {
      console.log('[WebRTC] Setting remote answer description...')
      await pc.setRemoteDescription(new RTCSessionDescription(answerSdp))
      this.updateState({ 
        hasRemoteDescription: true,
      })
      console.log('[WebRTC] Remote answer set. Flushing', this.state.pendingCandidates.length, 'candidates')
      await this.flushPendingCandidates()
    } catch (error) {
      console.error('[WebRTC] Failed to handle answer', error)
      this.updateState({ error: 'Không thể xử lý phản hồi cuộc gọi' })
    }
  }

  async handleIceCandidate(candidateData: RTCIceCandidateInit | string) {
    if (!this.state.pc || !candidateData) return

    try {
      const candidate = new RTCIceCandidate(
        typeof candidateData === 'string' ? { candidate: candidateData } : candidateData
      )

      // FIX BUG #10: Deduplicate ICE candidates to prevent network replay attacks
      // and avoid double-adding when the same candidate arrives via multiple event names.
      // Use sdpMid + sdpMLineIndex as the unique key.
      const candidateKey = `${candidate.sdpMid ?? ''}:${candidate.sdpMLineIndex ?? -1}:${candidate.credential ?? ''}`;
      if (this.processedIceCandidates.has(candidateKey)) {
        console.log('[WebRTC] Skipping duplicate ICE candidate:', candidateKey);
        return;
      }
      this.processedIceCandidates.add(candidateKey);

      if (!this.state.hasRemoteDescription) {
        console.log('[WebRTC] Queueing ICE candidate (remote description not ready)')
        this.updateState({ pendingCandidates: [...this.state.pendingCandidates, candidate] })
      } else {
        await this.state.pc.addIceCandidate(candidate)
        console.log('[WebRTC] ICE candidate added successfully')
      }
    } catch (error) {
      console.warn('[WebRTC] Failed to add ICE candidate', error)
    }
  }

  private async flushPendingCandidates() {
    if (!this.state.pc || this.state.pendingCandidates.length === 0) return

    for (const candidate of this.state.pendingCandidates) {
      try {
        await this.state.pc.addIceCandidate(candidate)
      } catch (error) {
        console.error('[WebRTC] Failed to flush ICE candidate', error)
      }
    }
    this.updateState({ pendingCandidates: [] })
  }

  private startRingTimeout() {
    this.stopRingTimeout()
    this.ringTimeoutTimer = setTimeout(() => {
      if (!this.state.isConnected && !this.state.isEnded) {
        this.endCall('no-answer-timeout')
      }
    }, this.RING_TIMEOUT_MS)
  }

  private stopRingTimeout() {
    if (this.ringTimeoutTimer) {
      clearTimeout(this.ringTimeoutTimer)
      this.ringTimeoutTimer = null
    }
  }

  endCall(reason: string = 'hangup', notifyPeer: boolean = true) {
    if (this.state.isEnded) return

    const safeReason = typeof reason === 'string' ? reason : 'hangup';
    console.log('[WebRTC] Ending call, reason:', safeReason)

    this.stopRingTimeout()

    this.updateState({ isEnded: true, isConnected: false })

    if (notifyPeer && this.socket) {
      const duration = this.state.startedAt ? Math.floor((Date.now() - this.state.startedAt) / 1000) : 0
      const outcome = this.state.isConnected ? 'completed' : (this.isCaller ? 'canceled' : 'missed')

      const endPayload = {
        conversationId: this.conversationId,
        callId: this.callId,
        senderUserId: this.currentUserId,
        targetUserId: this.peerUserId,
        callerId: this.isCaller ? this.currentUserId : this.peerUserId,
        calleeId: this.isCaller ? this.peerUserId : this.currentUserId,
        roomId: this.conversationId,
        reason: safeReason,
        duration: duration,
        outcome: outcome,
        startedAt: this.state.startedAt ?? null,
        direction: this.isCaller ? 'outgoing' : 'incoming',
      }
      this.socket.emit('call.end', endPayload)
      this.socket.emit('call:end', endPayload)
    }

    // Cleanup tracks
    this.state.localStream?.getTracks().forEach((track) => track.stop())
    this.state.remoteStream?.getTracks().forEach((track) => track.stop())

    // Close PC
    if (this.state.pc) {
      this.state.pc.close()
    }

    this.updateState({
      pc: null,
      localStream: null,
      remoteStream: null,
      pendingCandidates: [],
      hasRemoteDescription: false,
    });
    this.processedIceCandidates.clear();
  }

  toggleMic() {
    if (this.state.localStream) {
      const audioTracks = this.state.localStream.getAudioTracks()
      const newStatus = !this.state.isMicOn
      audioTracks.forEach(track => {
        track.enabled = newStatus
      })
      this.updateState({ isMicOn: newStatus })
    }
  }

  toggleCamera() {
    if (this.state.localStream && !this.audioOnly) {
      const videoTracks = this.state.localStream.getVideoTracks()
      const newStatus = !this.state.isCameraOn
      videoTracks.forEach(track => {
        track.enabled = newStatus
      })
      this.updateState({ isCameraOn: newStatus })
    }
  }
}
